src/main/java/de/dm/prom/structuredlogging/MdcTaskDecorator.java

```diff
@@ -27,34 +29,44 @@
      * will log a WARNing and not copy the MDC context if the other thread already has content in its MDC context.
      *
      * @param runnable runnable to run
+     * @param overwriteStrategy strategy to use when encountering MDC content in decorated threads
      *
      * @return the decorated runnable
      */
-    public static Runnable decorate(Runnable runnable) {
+    public static Runnable decorate(Runnable runnable, OverwriteStrategy overwriteStrategy) {
         Map<String, String> contextMap = MDC.getCopyOfContextMap();
         return () -> {
             boolean contextWasSet = false;
+            Map<String, String> threadsContextMap = MDC.getCopyOfContextMap();
             try {
-                if (hasContent(contextMap)) {
-                    Map<String, String> threadsContextMap = MDC.getCopyOfContextMap();
-                    if (!hasContent(threadsContextMap)) {
+                if (getKeys(contextMap).isPresent()) {
+                    Optional<Set<String>> presentKeys = getKeys(threadsContextMap);
+                    if (overwriteStrategy != OverwriteStrategy.PREVENT_OVERWRITE || !presentKeys.isPresent()) {
                         MDC.setContextMap(contextMap);
                         contextWasSet = true;
+
+                        if (overwriteStrategy == OverwriteStrategy.LOG_OVERWRITE) {
+                            log.warn("MDC context was set despite MDC keys being present in target thread. MDC keys present: {}", presentKeys.get());
```

Comment:
.get() will fail if presentKeys is not present. The if in line 44 is true if presentKeys is not present.

Reply:
Take a look at https://github.com/dm-drogeriemarkt/structured-logging/pull/8/commits/fc4801a19618be521c1e6721d9a85e32bda52cb3

Reply:
looks good.